### PR TITLE
MTDSA-8702 Added RequestParser + Validator for Submit VAT Details for Insolvent Trader

### PR DESCRIPTION
**Pull request by ngangaNjiraini**

### CODE REVIEW

File: app/v1/controllers/requestParsers/validators/SubmitReturnValidator.scala [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+      JsonFormatValidation.validate[SubmitReturnRequestBody](data.body.json)
+    )
+  }
+
+  private def bodyValueValidator: SubmitReturnRawData => List[List[MtdError]] = { data =>
+    val requestBodyData = data.body.json.as[SubmitReturnRequestBody]
+
+    List(Validator.flattenErrors(List(
+      PeriodKeyValidation.validate(requestBodyData.periodKey).map(
+        _.copy(paths = Some(Seq(s"/periodKey")))
```
Comment (james-work-account) [inline app/v1/controllers/requestParsers/validators/SubmitReturnValidator.scala]: Do you need a path here? There's only one periodKey, it's not like `ValueError` which is used for multiple fields.

File: app/v1/controllers/requestParsers/validators/validations/DecimalValueValidation.scala [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+  }
+
+  def validate(amount: BigDecimal,
+               maxScale: Int = 2,
+               minValue: BigDecimal = 0,
+               maxValue: BigDecimal = 9999999999999.99,
+               path: String,
+               message: String = BIG_DECIMAL_MINIMUM_INCLUSIVE): List[MtdError] = {
+
+    val scaleCheck = !(amount.scale > maxScale)
```
Comment (james-work-account) [inline app/v1/controllers/requestParsers/validators/validations/DecimalValueValidation.scala]: Can these Booleans be reversed? like `amount.scale <= maxScale` - having negative checks can be a bit confusing to understand when skimming.

File: app/v1/controllers/requestParsers/validators/validations/DecimalValueValidation.scala [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    val rangeCheck = !(amount > maxValue || amount < minValue)
```
Comment (james-work-account) [inline app/v1/controllers/requestParsers/validators/validations/DecimalValueValidation.scala]: Can these Booleans be reversed? like `amount >= minValue && amount <= maxValue` - having negative checks can be a bit confusing to understand when skimming.

File: app/v1/controllers/requestParsers/validators/validations/NonDecimalValueValidation.scala [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+object NonDecimalValueValidation extends ValueFormatErrorMessages {
+
+  def validate(amount: BigDecimal,
+               minValue: BigDecimal = -9999999999999.0,
+               maxValue: BigDecimal = 9999999999999.0,
+               path: String,
+               message: String = BIG_NONDECIMAL_MINIMUM_INCLUSIVE): List[MtdError] = {
+
+    val rangeCheck = !(amount > maxValue || amount < minValue)
```
Comment (james-work-account) [inline app/v1/controllers/requestParsers/validators/validations/NonDecimalValueValidation.scala]: Can these Booleans be reversed? like `amount >= minValue && amount <= maxValue` - having negative checks can be a bit confusing to understand when skimming.

File: app/v1/controllers/requestParsers/validators/validations/DecimalValueValidation.scala [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+    if (rangeCheck && scaleCheck) NoValidationErrors else List(
```
Comment (james-work-account) [inline app/v1/controllers/requestParsers/validators/validations/DecimalValueValidation.scala]: Please rename these Booleans to something more meaningful - like `amountIsInRange` and `amountScaleIsCorrect` or something similar.

File: app/v1/controllers/requestParsers/validators/validations/NonDecimalValueValidation.scala [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+    if (rangeCheck && amount.isWhole()) NoValidationErrors else List(
```
Comment (james-work-account) [inline app/v1/controllers/requestParsers/validators/validations/NonDecimalValueValidation.scala]: Please rename this Boolean to something more meaningful - like `amountIsInRange` or something similar.

File: app/v1/controllers/requestParsers/validators/SubmitReturnValidator.scala [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+      NonDecimalValueValidation.validate(
+        amount = requestBodyData.	totalValueGoodsSuppliedExVAT,
+        path = "/totalValueGoodsSuppliedExVAT"
+      ),
+      NonDecimalValueValidation.validate(
+        amount = requestBodyData.totalAcquisitionsExVAT,
+        path = "/totalAcquisitionsExVAT"
+      ),
+      DateFormatValidation.validate(requestBodyData.receivedAt).map(
+        _.copy(paths = Some(Seq(s"/receivedAt")))
```
Comment (james-work-account) [inline app/v1/controllers/requestParsers/validators/SubmitReturnValidator.scala]: Do you need a path here? There's only one receivedAt, it's not like ValueError which is used for multiple fields.

File: app/v1/controllers/requestParsers/validators/SubmitReturnValidator.scala [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+      ),
+      UniqueIDValidation.validate(requestBodyData.uniqueId).map(
+        _.copy(paths = Some(Seq(s"/uniqueId")))
```
Comment (james-work-account) [inline app/v1/controllers/requestParsers/validators/SubmitReturnValidator.scala]: Do you need a path here? There's only one uniqueId, it's not like ValueError which is used for multiple fields.

**Review by james-work-account — CHANGES_REQUESTED** [review]

File: app/v1/controllers/requestParsers/validators/SubmitReturnValidator.scala [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+import v1.models.request.{SubmitReturnRawData, SubmitReturnRequestBody}
+
+class SubmitReturnValidator extends Validator[SubmitReturnRawData] with ValueFormatErrorMessages{
+
+  private val validationSet = List(parameterFormatValidation, bodyValueValidator)
+
+  private def parameterFormatValidation: SubmitReturnRawData => List[List[MtdError]] = (data: SubmitReturnRawData) => {
+    List(
+      VrnValidation.validate(data.vrn),
+      JsonFormatValidation.validate[SubmitReturnRequestBody](data.body.json)
```
Comment (Taiwo-AB) [inline app/v1/controllers/requestParsers/validators/SubmitReturnValidator.scala:35]: This isn't part of the parameter validation. See [here](https://github.com/hmrc/individuals-income-received-api/blob/master/app/v1/controllers/requestParsers/validators/AmendDividendsValidator.scala#L49)

File: app/v1/controllers/requestParsers/validators/validations/DecimalValueValidation.scala [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+ * limitations under the License.
+ */
+
+package v1.controllers.requestParsers.validators.validations
+
+import v1.models.errors.{MtdError, ValueFormatError}
+
+object DecimalValueValidation extends ValueFormatErrorMessages {
+
+  def validateOptional(amount: Option[BigDecimal],
```
Comment (Taiwo-AB) [inline app/v1/controllers/requestParsers/validators/validations/DecimalValueValidation.scala]: I don't think this is being used as all fields are mandatory. Please remove

File: app/v1/controllers/requestParsers/validators/validations/PeriodKeyValidation.scala [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package v1.controllers.requestParsers.validators.validations
+
+import v1.models.errors.{MtdError, PeriodKeyFormatError}
+
+object PeriodKeyValidation {
+  private val vrnRegex = """(^[0-9]{2}[A-Z][A-Z0-9]$)|(^#[0-9]{3}$)|(^[0-9]{4}$)"""
```
Comment (Taiwo-AB) [inline app/v1/controllers/requestParsers/validators/validations/PeriodKeyValidation.scala]: this is not vrnRegex

File: app/v1/controllers/requestParsers/validators/validations/ValueFormatErrorsMessages.scala [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package v1.controllers.requestParsers.validators.validations
+
+protected[validators] trait ValueFormatErrorMessages {
+  val ZERO_MINIMUM_INCLUSIVE = "The field should be between 0 and 99999999999.99"
+  val ZERO_MINIMUM_INTEGER_INCLUSIVE = "The field should be between 0 and 99"
```
Comment (Taiwo-AB) [inline app/v1/controllers/requestParsers/validators/validations/ValueFormatErrorsMessages.scala]: This should be removed it is unused

File: app/v1/controllers/requestParsers/validators/validations/ValueFormatErrorsMessages.scala [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+  val ZERO_MINIMUM_BIG_INTEGER_INCLUSIVE = "The field should be 0 or more"
```
Comment (Taiwo-AB) [inline app/v1/controllers/requestParsers/validators/validations/ValueFormatErrorsMessages.scala]: This should be removed it is unused

File: app/v1/controllers/requestParsers/validators/validations/ValueFormatErrorsMessages.scala [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package v1.controllers.requestParsers.validators.validations
+
+protected[validators] trait ValueFormatErrorMessages {
+  val ZERO_MINIMUM_INCLUSIVE = "The field should be between 0 and 99999999999.99"
```
Comment (Taiwo-AB) [inline app/v1/controllers/requestParsers/validators/validations/ValueFormatErrorsMessages.scala]: Can this be updated to `The field should be between 0.00 and 99999999999.99` match the netVatDue description in the doc. Thanks

File: test/v1/controllers/requestParsers/validators/SubmitReturnValidatorSpec.scala [hunk 15]
```diff
@@ -0,0 +1,285 @@
...
+  private val validRawRequestBody = AnyContentAsJson(validRequestBodyJson)
+  private val emptyRawRequestBody = AnyContentAsJson(emptyRequestBodyJson)
+  private val nonValidRawRequestBody = AnyContentAsJson(nonValidRequestBodyJson)
+  private val invalidPeriodKeyRequestBody = AnyContentAsJson(invalidPeriodKeyRequestBodyJson)
+  private val invalidReceivedAtRequestBody = AnyContentAsJson(invalidReceivedAtRequestBodyJson)
+  private val invalidUniqueIdRequestBody = AnyContentAsJson(invalidUniqueIdRequestBodyJson)
+  private val invalidVatDueSalesRequestBody = AnyContentAsJson(invalidVatDueSalesRequestBodyJson)
+  private val allInvalidValueRawRequestBody = AnyContentAsJson(allInvalidValueRequestBodyJson)
+
+  class Test extends MockAppConfig {
```
Comment (Taiwo-AB) [inline test/v1/controllers/requestParsers/validators/SubmitReturnValidatorSpec.scala]: I don't think you need all this. Are you using the App config?. I think you can just use this 
`val validator: SubmitReturnValidator = new SubmitReturnValidator()`

File: test/v1/controllers/requestParsers/validators/validations/DateTimeFormatValidationSpec.scala [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+ */
+
+package v1.controllers.requestParsers.validators.validations
+
+import support.UnitSpec
+import v1.models.errors.ReceivedAtFormatError
+
+class DateTimeFormatValidationSpec extends UnitSpec {
+
+  "DateFormatValidation" when {
```
Comment (Taiwo-AB) [inline test/v1/controllers/requestParsers/validators/validations/DateTimeFormatValidationSpec.scala]: Update this to DateTimeFormatValidation

File: app/v1/controllers/requestParsers/validators/validations/DateTimeFormatValidation.scala [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+import v1.models.errors.{MtdError, ReceivedAtFormatError}
+
+import scala.util.{Failure, Success, Try}
+
+
+object DateTimeFormatValidation {
+
+  val dateTimeFormat: DateTimeFormatter = DateTimeFormatter.ofPattern("uuuu-MM-dd'T'HH:mm:ss'Z'").withResolverStyle(ResolverStyle.STRICT)
+
+  def validate(date: String): List[MtdError] = Try {
```
Comment (Taiwo-AB) [inline app/v1/controllers/requestParsers/validators/validations/DateTimeFormatValidation.scala]: The format is in date-time can this be updated to `def validate(dateTime: String)`. The test will need updating too

File: test/v1/controllers/requestParsers/SubmitReturnRequestParserSpec.scala [hunk 18]
```diff
@@ -0,0 +1,161 @@
...
+          ValueFormatError.copy(
+            paths = Some(List(
+              "/totalValueSalesExVAT",
+              "/totalValuePurchasesExVAT",
+              "/totalValueGoodsSuppliedExVAT",
+              "/totalAcquisitionsExVAT"
+            )),
+            message = "The field should be between -9999999999999.99 and 9999999999999.99"
+          ),
+          ReceivedAtFormatError.copy(
```
Comment (Taiwo-AB) [inline test/v1/controllers/requestParsers/SubmitReturnRequestParserSpec.scala]: This shouldn't have paths

File: test/v1/controllers/requestParsers/validators/SubmitReturnValidatorSpec.scala [hunk 19]
```diff
@@ -0,0 +1,273 @@
...
+  }
+
+  "running a validation" should {
+    "return no errors" when {
+      "a valid request is supplied" in new Test {
+        validator.validate(SubmitReturnRawData(validVrn, validRawRequestBody)) shouldBe Nil
+      }
+    }
+
+    "return VrnFormatError errors" when {
```
Comment (Taiwo-AB) [inline test/v1/controllers/requestParsers/validators/SubmitReturnValidatorSpec.scala]: remove s at the end

File: test/v1/controllers/requestParsers/SubmitReturnRequestParserSpec.scala [hunk 20]
```diff
@@ -0,0 +1,161 @@
...
+            paths = Some(List("/receivedAt"))
+          ),
+          UniqueIDFormatError.copy(
```
Comment (Taiwo-AB) [inline test/v1/controllers/requestParsers/SubmitReturnRequestParserSpec.scala]: This shouldn't have paths

File: test/v1/controllers/requestParsers/SubmitReturnRequestParserSpec.scala [hunk 21]
```diff
@@ -0,0 +1,161 @@
...
+            |  "receivedAt": "time",
+            |  "uniqueId": "0123456789123"
+            |}
+          """.stripMargin
+        )
+
+        private val allInvalidValueRawRequestBody = AnyContentAsJson(allInvalidValueRequestBodyJson)
+
+        private val allInvalidValueErrors = List(
+          PeriodKeyFormatError.copy(
```
Comment (Taiwo-AB) [inline test/v1/controllers/requestParsers/SubmitReturnRequestParserSpec.scala]: This shouldn't use paths

File: test/v1/controllers/requestParsers/validators/validations/DateTimeFormatValidationSpec.scala [hunk 22]
```diff
@@ -0,0 +1,40 @@
...
+    "validate" must {
+      "return an empty list for a valid date" in {
```
Comment (Taiwo-AB) [inline test/v1/controllers/requestParsers/validators/validations/DateTimeFormatValidationSpec.scala]: The wordings should be also updated as a valid date will not return an empty list

File: test/v1/controllers/requestParsers/validators/validations/DateTimeFormatValidationSpec.scala [hunk 23]
```diff
@@ -0,0 +1,40 @@
...
+        DateTimeFormatValidation.validate(
+          date = "2020-05-05T12:01:00Z"
+        ) shouldBe NoValidationErrors
+      }
+
+      "return a ReceivedAtFormatError for an invalid date" in {
```
Comment (Taiwo-AB) [inline test/v1/controllers/requestParsers/validators/validations/DateTimeFormatValidationSpec.scala]: Same here

File: test/v1/controllers/requestParsers/SubmitReturnRequestParserSpec.scala [hunk 24]
```diff
@@ -0,0 +1,161 @@
...
+            message = "The field should be between 0.00 and 99999999999.99"
+          ),
+          ValueFormatError.copy(
+            paths = Some(List(
+              "/totalValueSalesExVAT",
+              "/totalValuePurchasesExVAT",
+              "/totalValueGoodsSuppliedExVAT",
+              "/totalAcquisitionsExVAT"
+            )),
+            message = "The field should be between -9999999999999.99 and 9999999999999.99"
```
Comment (Taiwo-AB) [inline test/v1/controllers/requestParsers/SubmitReturnRequestParserSpec.scala]: This isn't the right message for the non decimal fields

File: app/v1/controllers/requestParsers/validators/validations/DateTimeFormatValidation.scala [hunk 25]
```diff
@@ -0,0 +1,37 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package v1.controllers.requestParsers.validators.validations
+
+import java.time.{LocalDateTime}
```
Comment (Taiwo-AB) [inline app/v1/controllers/requestParsers/validators/validations/DateTimeFormatValidation.scala]: remove {}

File: app/v1/controllers/requestParsers/validators/validations/ValueFormatErrorsMessages.scala [hunk 26]
```diff
@@ -0,0 +1,23 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package v1.controllers.requestParsers.validators.validations
+
+protected[validators] trait ValueFormatErrorMessages {
```
Comment (Taiwo-AB) [inline app/v1/controllers/requestParsers/validators/validations/ValueFormatErrorsMessages.scala]: Update file name to match this `ValueFormatErrorMessages` on this line - remove s in Errors in /ValueFormatErrorsMessages.scala

**Review by Taiwo-AB — APPROVED** [review]

**Pull request merged**